### PR TITLE
[Logstash]: Http client - v0.4.0

### PR DESCRIPTION
**Pull request by gaetan-deputier**
### What does this PR do?

- Support by default HTTP forwarding
- Support HTTP compression by default
- Add option to disable SSL validation for TCP & HTTP
- Add some basic config & logic tests

### CODE REVIEW

File: lib/logstash/outputs/datadog_logs.rb [hunk 1]
```diff
@@ -6,67 +6,193 @@
...
+
+    def initialize(logger, use_ssl, no_ssl_validation, host, port, use_compression, api_key)
+      @logger = logger
+      protocol = use_ssl ? "https" : "http"
+      @url = "#{protocol}://#{host}:#{port.to_s}/v1/input/#{api_key}"
+      @headers = {"Content-Type" => "application/json"}
+      if use_compression
+        @headers["Content-Encoding"] = "gzip"
+      end
+      logger.info("Starting HTTP connection to #{protocol}://#{@host}:#{port.to_s}")
```
Comment (blemale) [inline lib/logstash/outputs/datadog_logs.rb]: You should probably also log the use or not of the compression

File: lib/logstash/outputs/datadog_logs.rb [hunk 2]
```diff
@@ -6,67 +6,193 @@
...
+      return payload
+    end
+  end
+
+  # Group events in batches
+  def batch_events(events, max_batch_size)
+    batches = []
+    current_batch = []
+    events.each_with_index do |event, i|
+      if i > 0 and i % max_batch_size == 0
```
Comment (blemale) [inline lib/logstash/outputs/datadog_logs.rb]: We should also check the max payload size in bytes, isn't it?

File: lib/logstash/outputs/datadog_logs.rb [hunk 3]
```diff
@@ -6,67 +6,193 @@
...
+    return batches
+  end
+
+  # Compress logs with GZIP
+  def gzip_compress(payload, compression_level)
+    gz = StringIO.new
+    gz.set_encoding("BINARY")
+    z = Zlib::GzipWriter.new(gz, compression_level)
+    z.write(payload)
+    z.close
```
[the remark this answers is not in the document]
Comment (gaetan-deputier) [reply lib/logstash/outputs/datadog_logs.rb]: Correct, updated!

File: lib/logstash/outputs/datadog_logs.rb (hunk 1 above)
Comment (gaetan-deputier) [reply lib/logstash/outputs/datadog_logs.rb]: Added

File: lib/logstash/outputs/datadog_logs.rb (hunk 2 above)
Comment (gaetan-deputier) [reply lib/logstash/outputs/datadog_logs.rb]: Correct, updated. I've reworked a bit the code to ensure we serialize events before computing the size.

File: lib/logstash/outputs/datadog_logs.rb [hunk 4]
```diff
@@ -6,67 +6,193 @@
 # encoding: utf-8
 require "logstash/outputs/base"
 require "logstash/namespace"
+require "zlib"
```
[the remark this answers is not in the document]
Comment (gaetan-deputier) [reply lib/logstash/outputs/datadog_logs.rb]: Yes it's required. ZLib's module is of ruby's standard library so I'd expect ruby installation to break in this case). 
There is a pure ruby implementation (pr-ruby) but no updates for more than a year.
I'll mention it in the Readme though

File: lib/logstash/outputs/datadog_logs.rb [hunk 5]
```diff
@@ -6,67 +6,231 @@
...
+    "[#{batched_events.join(',')}]"
+  end
+
+  # Group HTTP events in batches
+  def batch_http_events(encoded_events, max_batch_length, max_request_size)
+    batches = []
+    current_batch = []
+    current_batch_size = 0
+    encoded_events.each_with_index do |event, i|
+      encoded_event = event.last
```
Comment (blemale) [inline lib/logstash/outputs/datadog_logs.rb]: I'm not sure to understand this line, why `last` returns the `encoded_event`?

Comment (gaetan-deputier) [reply lib/logstash/outputs/datadog_logs.rb]: Good question. Logstash returns both the original event and the encoded event after being encoded with @codec. So last is here to get the encoded event.

**Pull request merged**